[
  {
    "package": {
      "name": "pygments",
      "version": "2.5.2",
      "ecosystem": "PyPI"
    },
    "vulnerabilities": [
      {
        "modified": "2023-08-07T11:31:32Z",
        "published": "2021-04-20T16:35:47Z",
        "schema_version": "1.4.0",
        "id": "GHSA-9w8r-397f-prfh",
        "aliases": [
          "CVE-2021-20270"
        ],
        "summary": "Infinite Loop in Pygments",
        "details": "An infinite loop in SMLLexer in Pygments versions 1.5 to 2.7.3 may lead to denial of service when performing syntax highlighting of a Standard ML (SML) source file, as demonstrated by input that only contains the \"exception\" keyword.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "pygments",
              "purl": "pkg:pypi/pygments"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "1.5"
                  },
                  {
                    "fixed": "2.7.4"
                  }
                ]
              }
            ],
            "versions": [
              "1.5",
              "1.6",
              "1.6rc1",
              "2.0",
              "2.0.1",
              "2.0.2",
              "2.0rc1",
              "2.1",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.2.0",
              "2.3.0",
              "2.3.1",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.5.1",
              "2.5.2",
              "2.6.0",
              "2.6.1",
              "2.7.0",
              "2.7.1",
              "2.7.2",
              "2.7.3"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2021/04/GHSA-9w8r-397f-prfh/GHSA-9w8r-397f-prfh.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                ""
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H"
          }
        ],
        "references": [
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2021-20270"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pygments/pygments/commit/f91804ff4772e3ab41f46e28d370f57898700333"
          },
          {
            "type": "WEB",
            "url": "https://bugzilla.redhat.com/show_bug.cgi?id=1922136"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/pygments/pygments"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00003.html"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00006.html"
          },
          {
            "type": "WEB",
            "url": "https://www.debian.org/security/2021/dsa-4889"
          },
          {
            "type": "WEB",
            "url": "https://www.oracle.com/security-alerts/cpuoct2021.html"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-835"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2021-04-12T20:58:03Z",
          "nvd_published_at": "2021-03-23T17:15:00Z",
          "severity": "HIGH"
        }
      },
      {
        "modified": "2023-07-26T21:46:26Z",
        "published": "2023-07-19T15:30:26Z",
        "schema_version": "1.4.0",
        "id": "GHSA-mrwq-x4v8-fh7p",
        "aliases": [
          "CVE-2022-40896"
        ],
        "summary": "Pygments vulnerable to ReDoS",
        "details": "A ReDoS issue was discovered in `pygments/lexers/smithy.py` in Pygments until 2.15.0 via SmithyLexer.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "pygments",
              "purl": "pkg:pypi/pygments"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.15.0"
                  }
                ]
              }
            ],
            "versions": [
              "0.10",
              "0.11",
              "0.11.1",
              "0.5",
              "0.5.1",
              "0.6",
              "0.7",
              "0.7.1",
              "0.8",
              "0.8.1",
              "0.9",
              "1.0",
              "1.1",
              "1.1.1",
              "1.2",
              "1.2.1",
              "1.2.2",
              "1.3",
              "1.3.1",
              "1.4",
              "1.5",
              "1.6",
              "1.6rc1",
              "2.0",
              "2.0.1",
              "2.0.2",
              "2.0rc1",
              "2.1",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.10.0",
              "2.11.0",
              "2.11.1",
              "2.11.2",
              "2.12.0",
              "2.13.0",
              "2.14.0",
              "2.2.0",
              "2.3.0",
              "2.3.1",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.5.1",
              "2.5.2",
              "2.6.0",
              "2.6.1",
              "2.7.0",
              "2.7.1",
              "2.7.2",
              "2.7.3",
              "2.7.4",
              "2.8.0",
              "2.8.1",
              "2.9.0"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/07/GHSA-mrwq-x4v8-fh7p/GHSA-mrwq-x4v8-fh7p.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "pygments.lexers.templates.SqlJinjaLexer.analyse_text",
                "pygments.lexers.smithy.SmithyLexer",
                "pygments.lexers.configs.PropertiesLexer"
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:U/C:N/I:N/A:H"
          }
        ],
        "references": [
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2022-40896"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pygments/pygments/commit/97eb3d5ec7c1b3ea4fcf9dee30a2309cf92bd194"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pygments/pygments/commit/dd52102c38ebe78cd57748e09f38929fd283ad04"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pygments/pygments/commit/fdf182a7af85b1deeeb637ca970d31935e7c9d52"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/pygments/pygments"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pygments/pygments/blob/master/pygments/lexers/smithy.py#L61"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pypa/advisory-database/tree/main/vulns/pygments/PYSEC-2023-117.yaml"
          },
          {
            "type": "WEB",
            "url": "https://pypi.org/project/Pygments/"
          },
          {
            "type": "WEB",
            "url": "https://pyup.io/posts/pyup-discovers-redos-vulnerabilities-in-top-python-packages-part-2/"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-434"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2023-07-20T11:00:04Z",
          "nvd_published_at": null,
          "severity": "MODERATE"
        }
      },
      {
        "modified": "2023-08-07T11:45:47Z",
        "published": "2021-03-29T16:33:03Z",
        "schema_version": "1.4.0",
        "id": "GHSA-pq64-v7f5-gqh8",
        "aliases": [
          "CVE-2021-27291"
        ],
        "summary": "Pygments vulnerable to Regular Expression Denial of Service (ReDoS)",
        "details": "In pygments 1.1+, fixed in 2.7.4, the lexers used to parse programming languages rely heavily on regular expressions. Some of the regular expressions have exponential or cubic worst-case complexity and are vulnerable to ReDoS. By crafting malicious input, an attacker can cause a denial of service.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "pygments",
              "purl": "pkg:pypi/pygments"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "1.1"
                  },
                  {
                    "fixed": "2.7.4"
                  }
                ]
              }
            ],
            "versions": [
              "1.1",
              "1.1.1",
              "1.2",
              "1.2.1",
              "1.2.2",
              "1.3",
              "1.3.1",
              "1.4",
              "1.5",
              "1.6",
              "1.6rc1",
              "2.0",
              "2.0.1",
              "2.0.2",
              "2.0rc1",
              "2.1",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.2.0",
              "2.3.0",
              "2.3.1",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.5.1",
              "2.5.2",
              "2.6.0",
              "2.6.1",
              "2.7.0",
              "2.7.1",
              "2.7.2",
              "2.7.3"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2021/03/GHSA-pq64-v7f5-gqh8/GHSA-pq64-v7f5-gqh8.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                ""
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H"
          }
        ],
        "references": [
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2021-27291"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pygments/pygments/commit/2e7e8c4a7b318f4032493773732754e418279a14"
          },
          {
            "type": "WEB",
            "url": "https://gist.github.com/b-c-ds/b1a2cc0c68a35c57188575eb496de5ce"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/pygments/pygments"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/03/msg00024.html"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00003.html"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00006.html"
          },
          {
            "type": "WEB",
            "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/GSJRFHALQ7E3UV4FFMFU2YQ6LUDHAI55/"
          },
          {
            "type": "WEB",
            "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/WSLD67LFGXOX2K5YNESSWAS4AGZIJTUQ/"
          },
          {
            "type": "WEB",
            "url": "https://www.debian.org/security/2021/dsa-4878"
          },
          {
            "type": "WEB",
            "url": "https://www.debian.org/security/2021/dsa-4889"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-400"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2021-03-22T20:28:40Z",
          "nvd_published_at": "2021-03-17T13:15:00Z",
          "severity": "HIGH"
        }
      },
      {
        "modified": "2021-08-27T03:22:17Z",
        "published": "2021-03-23T17:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2021-140",
        "aliases": [
          "CVE-2021-20270",
          "GHSA-9w8r-397f-prfh"
        ],
        "details": "An infinite loop in SMLLexer in Pygments versions 1.5 to 2.7.3 may lead to denial of service when performing syntax highlighting of a Standard ML (SML) source file, as demonstrated by input that only contains the \"exception\" keyword.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "pygments",
              "purl": "pkg:pypi/pygments"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "1.5"
                  },
                  {
                    "fixed": "2.7.4"
                  }
                ]
              }
            ],
            "versions": [
              "1.5",
              "1.6",
              "1.6rc1",
              "2.0",
              "2.0.1",
              "2.0.2",
              "2.0rc1",
              "2.1",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.2.0",
              "2.3.0",
              "2.3.1",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.5.1",
              "2.5.2",
              "2.6.0",
              "2.6.1",
              "2.7.0",
              "2.7.1",
              "2.7.2",
              "2.7.3"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/pygments/PYSEC-2021-140.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "REPORT",
            "url": "https://bugzilla.redhat.com/show_bug.cgi?id=1922136"
          },
          {
            "type": "ADVISORY",
            "url": "https://www.debian.org/security/2021/dsa-4889"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00003.html"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00006.html"
          },
          {
            "type": "ADVISORY",
            "url": "https://github.com/advisories/GHSA-9w8r-397f-prfh"
          }
        ]
      },
      {
        "modified": "2021-08-27T03:22:17Z",
        "published": "2021-03-17T13:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2021-141",
        "aliases": [
          "CVE-2021-27291",
          "GHSA-pq64-v7f5-gqh8"
        ],
        "details": "In pygments 1.1+, fixed in 2.7.4, the lexers used to parse programming languages rely heavily on regular expressions. Some of the regular expressions have exponential or cubic worst-case complexity and are vulnerable to ReDoS. By crafting malicious input, an attacker can cause a denial of service.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "pygments",
              "purl": "pkg:pypi/pygments"
            },
            "ranges": [
              {
                "type": "GIT",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2e7e8c4a7b318f4032493773732754e418279a14"
                  }
                ],
                "repo": "https://github.com/pygments/pygments"
              },
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "1.1"
                  },
                  {
                    "fixed": "2.7.4"
                  }
                ]
              }
            ],
            "versions": [
              "1.1",
              "1.1.1",
              "1.2",
              "1.2.1",
              "1.2.2",
              "1.3",
              "1.3.1",
              "1.4",
              "1.5",
              "1.6",
              "1.6rc1",
              "2.0",
              "2.0.1",
              "2.0.2",
              "2.0rc1",
              "2.1",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.2.0",
              "2.3.0",
              "2.3.1",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.5.1",
              "2.5.2",
              "2.6.0",
              "2.6.1",
              "2.7.0",
              "2.7.1",
              "2.7.2",
              "2.7.3"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/pygments/PYSEC-2021-141.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://gist.github.com/b-c-ds/b1a2cc0c68a35c57188575eb496de5ce"
          },
          {
            "type": "FIX",
            "url": "https://github.com/pygments/pygments/commit/2e7e8c4a7b318f4032493773732754e418279a14"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/03/msg00024.html"
          },
          {
            "type": "ADVISORY",
            "url": "https://www.debian.org/security/2021/dsa-4878"
          },
          {
            "type": "ADVISORY",
            "url": "https://www.debian.org/security/2021/dsa-4889"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00003.html"
          },
          {
            "type": "WEB",
            "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/WSLD67LFGXOX2K5YNESSWAS4AGZIJTUQ/"
          },
          {
            "type": "WEB",
            "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/GSJRFHALQ7E3UV4FFMFU2YQ6LUDHAI55/"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2021/05/msg00006.html"
          },
          {
            "type": "ADVISORY",
            "url": "https://github.com/advisories/GHSA-pq64-v7f5-gqh8"
          }
        ]
      },
      {
        "modified": "2023-07-19T17:26:16Z",
        "published": "2023-07-19T15:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2023-117",
        "aliases": [
          "CVE-2022-40896"
        ],
        "details": "A ReDoS issue was discovered in pygments/lexers/smithy.py in pygments through 2.15.0 via SmithyLexer.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "pygments",
              "purl": "pkg:pypi/pygments"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.15.1"
                  }
                ]
              }
            ],
            "versions": [
              "0.10",
              "0.11",
              "0.11.1",
              "0.5",
              "0.5.1",
              "0.6",
              "0.7",
              "0.7.1",
              "0.8",
              "0.8.1",
              "0.9",
              "1.0",
              "1.1",
              "1.1.1",
              "1.2",
              "1.2.1",
              "1.2.2",
              "1.3",
              "1.3.1",
              "1.4",
              "1.5",
              "1.6",
              "1.6rc1",
              "2.0",
              "2.0.1",
              "2.0.2",
              "2.0rc1",
              "2.1",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.10.0",
              "2.11.0",
              "2.11.1",
              "2.11.2",
              "2.12.0",
              "2.13.0",
              "2.14.0",
              "2.15.0",
              "2.2.0",
              "2.3.0",
              "2.3.1",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.5.1",
              "2.5.2",
              "2.6.0",
              "2.6.1",
              "2.7.0",
              "2.7.1",
              "2.7.2",
              "2.7.3",
              "2.7.4",
              "2.8.0",
              "2.8.1",
              "2.9.0"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/pygments/PYSEC-2023-117.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "PACKAGE",
            "url": "https://pypi.org/project/Pygments/"
          },
          {
            "type": "WEB",
            "url": "https://pyup.io/posts/pyup-discovers-redos-vulnerabilities-in-top-python-packages-part-2/"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pygments/pygments/blob/master/pygments/lexers/smithy.py#L61"
          }
        ]
      }
    ],
    "groups": [
      {
        "ids": [
          "GHSA-9w8r-397f-prfh",
          "PYSEC-2021-140"
        ]
      },
      {
        "ids": [
          "GHSA-mrwq-x4v8-fh7p",
          "PYSEC-2023-117"
        ]
      },
      {
        "ids": [
          "GHSA-pq64-v7f5-gqh8",
          "PYSEC-2021-141"
        ]
      }
    ]
  }
]